packages/itwinui-react/src/core/Select/Select.tsx

```diff
@@ -298,11 +302,22 @@ export const Select = React.forwardRef(
             if (option.disabled) {
               return;
             }
-            if (isSingleOnChange(onChange, multiple)) {
-              onChange?.(option.value);
+
+            // update internal value state and also call external onChange
+            if (isSingleOnChange(onChangeRef.current, multiple)) {
+              setUncontrolledValue(option.value);
+              onChangeRef.current?.(option.value);
               hide();
             } else {
-              onChange?.(option.value, isSelected ? 'removed' : 'added');
+              setUncontrolledValue((prev: T[]) =>
+                isSelected
+                  ? prev?.filter((i) => option.value !== i)
```

Comment:
Since prev is of type `T[]`, is the `?` needed?


```suggestion
                  ? prev.filter((i) => option.value !== i)
```

Reply:
you should never trust typescript. `uncontrolledValue` being initialized with `undefined`, so it may still be possible (can't hurt if not)

Reply:
Then would adding `undefined` as a possible type on L312 be better?

```diff
- setUncontrolledValue((prev: T[]) =>
+ setUncontrolledValue((prev: T[] | undefined) =>
```

Reply:
it's fine, the runtime JS code is unaffected and user-facing TS types are also unaffected.

Reply:
nit: That's true, but I mean if we add `undefined` to the possible types of `T`, then typescript would explicitly give an error when we don't add the `?` when accessing properties of T.

But without adding `undefined`, we have to remember that `T` can be `undefined` and safeguard accordingly.